### ISSUE TITLE
stack() on dask array produces inefficient chunking

### ISSUE THREAD

**Issue by crusaderky** [pydata/xarray#926]
Whe the stack() method is used on a xarray with dask backend, one would expect that every output chunk is produced by exactly 1 input chunk.

This is not the case, as stack() actually produces an extremely fragmented dask array:
https://gist.github.com/crusaderky/07991681d49117bfbef7a8870e3cba67

**Comment by shoyer** [comment]
Yes, this is unfortunate, but dask.array can't efficiently flatten (in C order) along chunked dimensions (see https://github.com/dask/dask/pull/758 for discussion).

One thing we *could* do is add an option to dask to flatten arrays in "block contiguous" order. Then we could add an option (or maybe even change the default) to xarray' stack to do block contiguous stacking. If you're interested in working on this, the place to start is the code for dask.array's ravel operation (see the link above for details).

**Comment by shoyer** [comment]
Given that we will need to be aware of the chunks layout in order to stack correctly by block, it might make sense to actually put all this logic in xarray, by using `dask.array.map_blocks` with `np.reshape`.

I'm not sure when I'll have the chance to work on this, so it might be a good relatively self-contained project for a new contributor.
